velox/docs/develop/timestamp.rst

```diff
@@ -0,0 +1,301 @@
+prior to that. 
+
+However, note that the physical representation of the timestamp is orthogonal
+to its logical meaning. For example, the timestamp represented by the ``0``
+integer was perceived at different absolute points in time depending on the
+observer’s timezone, and does not necessarily imply that it was observed in the
+UTC timezone. When a timestamp represents the number of seconds in UTC
+specifically (at that exact absolute instant in time), it may be called a *unix
+epoch* or *unix time*.
+
+Velox Classes and APIs
+----------------------
+
+Velox provides a few classes and APIs to allow developers to store, process,
+and convert timestamps across timezones:
+
+**Timestamp:** In Velox, timestamps are represented by the `Timestamp
+<https://github.com/facebookincubator/velox/blob/main/velox/type/Timestamp.h>`_
+class. The Timestamp class stores two 64 bit integers, one containing the
+number of seconds from ``1970-01-01 00:00:00``, and one containing the
+nanoseconds offset in that particular second, in order to provide nanosecond
+precision. A few more observations:
+
+* While “seconds” can be negative to represent time before ``1970-01-01
+  00:00:00``, “nanoseconds” are always positive.
+
+* Engines like Presto and Spark only support milliseconds precision, so
+  naturally “nanoseconds” are zeroed.
+
+* The Timestamp class only offers a physical representation of timestamps, but
+  does not carry logical information about its timezone. In other words, it
+  cannot, by itself, represent an absolute point in time.
+
+**Timezone IDs:** To physically represent timezones, Velox provides the
+`TimezoneMap.h <https://github.com/facebookincubator/velox/blob/main/velox/type/tz/TimeZoneMap.h>`_
+API. This API provides a 1:1 mapping from each available timezone to a
+monotonically increasing integer (a timezone ID), such that this integer can be
+used to efficiently represent timezones, preventing the use of inefficient
+timezone string names like ``America/Los_Angeles``. Considering there are about
+2k valid timezone definitions, 12 bits are enough to represent timezone IDs. 
+
+Timezone IDs in Velox are based on the id map used by Presto and are 
+`available here <https://github.com/prestodb/presto/blob/master/presto-common/src/main/resources/com/facebook/presto/common/type/zone-index.properties>`_. 
+They are automatically generated using `this script <https://github.com/facebookincubator/velox/blob/main/velox/type/tz/gen_timezone_database.py>`_. 
+While timezone IDs are an implementation detail and ideally should not leak
+outside of Velox execution, they are exposed if data containing
+TimestampWithTimezones are serialized, for example.
+
+**TimestampWithTimezone:** To represent an absolute point in time, Velox provides
+the `TimestampWithTimezone class <https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/types/TimestampWithTimeZoneType.h>`_.
```

Comment:
I don't think there is a TimestampWithTimezone class.

Reply:
well, it's an alias:

https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/types/TimestampWithTimeZoneType.h

I'll rephrase.